### PR TITLE
[ISSUES #8622] Add NacosEnvironment and add some unit tests

### PR DESCRIPTION
**Pull request by onewe**
- remove magic number
- refactor all of them
- add some unit tests for env
- modify search rule and reformat
- fix search order
- fix p3c error

This pr is part of the https://github.com/alibaba/nacos/issues/8622.

Please do not create a Pull Request without creating an issue first.

## What is the purpose of the change

XXXXX

## Brief changelog

XX

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

* [x] Make sure there is a Github issue filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
* [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
* [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
* [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/alibaba/nacos/tree/master/test).
* [x] Run `mvn -B clean package apache-rat:check findbugs:findbugs -Dmaven.test.skip=true` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.

### CODE REVIEW

File: client/src/main/java/com/alibaba/nacos/client/env/AbstractPropertySource.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.alibaba.nacos.client.env;
+
+abstract class AbstractPropertySource {
```
Comment (KomachiSion) [inline client/src/main/java/com/alibaba/nacos/client/env/AbstractPropertySource.java:19]: default access specifier is enough?

File: client/src/main/java/com/alibaba/nacos/client/env/DefaultSettingPropertySource.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+import java.io.InputStream;
+import java.net.URL;
+import java.util.Properties;
+
+class DefaultSettingPropertySource extends AbstractPropertySource {
+    
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultSettingPropertySource.class);
+    
+    private static final String DEFAULT_SETTING_PATH = "classpath:default_setting.properties";
```
Comment (KomachiSion) [inline client/src/main/java/com/alibaba/nacos/client/env/DefaultSettingPropertySource.java:31]: add `nacos` word to reduce the possible rate for conflict with other tools.

File: client/src/main/java/com/alibaba/nacos/client/env/PropertySourceSearch.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        
+        if (StringUtils.isBlank(pattern)) {
+            return createPropertySourceSearchWithDefaultOrder(propertySources);
+        }
+        
+        try {
+            final SourceType sourceType = SourceType.valueOf(pattern.toUpperCase());
+            return createPropertySourceSearchByFirstType(sourceType, propertySources);
+        } catch (Exception e) {
+            LOGGER.warn("first source type parse error, it will be use default order!");
```
Comment (KomachiSion) [inline client/src/main/java/com/alibaba/nacos/client/env/PropertySourceSearch.java:54]: print the error source type

File: client/src/main/java/com/alibaba/nacos/client/env/PropertySourceSearch.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                .collect(() -> tempList, List::add, List::addAll).stream().map(sourceMap::get)
+                .collect(Collectors.toList());
+        
+        return new PropertySourceSearch(collect);
+    }
+    
+    <T> T search(Function<AbstractPropertySource, String> function, Supplier<String> elseGet, Function<String, T> convert) {
+        String ret;
+        for (AbstractPropertySource propertySource : propertySources) {
+            ret = function.apply(propertySource);
```
Comment (KomachiSion) [inline client/src/main/java/com/alibaba/nacos/client/env/PropertySourceSearch.java:84]: Why not directly call `AbstractPropertySource.getProperty`?

**Review by KomachiSion — CHANGES_REQUESTED** [review]

File: client/src/main/java/com/alibaba/nacos/client/env/AbstractPropertySource.java (hunk 1 above)
Comment (onewe) [reply client/src/main/java/com/alibaba/nacos/client/env/AbstractPropertySource.java:19]: That's enough and don't need to expose some details to another.

File: client/src/main/java/com/alibaba/nacos/client/env/PropertySourceSearch.java (hunk 4 above)
Comment (onewe) [reply client/src/main/java/com/alibaba/nacos/client/env/PropertySourceSearch.java]: I not't only have corrected it but add some converters as well.

File: client/src/main/java/com/alibaba/nacos/client/env/PropertySourceSearch.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            if (ret != null) {
+                return convert.apply(ret);
+            }
+        }
+        
+        ret = elseGet.get();
```
[the remark this answers is not in the document]
Comment (onewe) [reply client/src/main/java/com/alibaba/nacos/client/env/PropertySourceSearch.java]: Got it

File: client/src/main/java/com/alibaba/nacos/client/env/SystemEnvPropertySource.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        this.env = System.getenv();
+    }
+    
+    @Override
+    SourceType getType() {
+        return SourceType.SYS;
+    }
+    
+    @Override
+    String getProperty(String key) {
```
Comment (KomachiSion) [inline client/src/main/java/com/alibaba/nacos/client/env/SystemEnvPropertySource.java:35]: should we try original case first, if not match, then try upperCase?

Comment (KomachiSion) [reply client/src/main/java/com/alibaba/nacos/client/env/SystemEnvPropertySource.java:35]: How about spring read env properties?

File: client/src/main/java/com/alibaba/nacos/client/env/DefaultSettingPropertySource.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+import java.io.InputStream;
+import java.net.URL;
+import java.util.Properties;
+
+class DefaultSettingPropertySource extends AbstractPropertySource {
+    
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultSettingPropertySource.class);
+    
+    private static final String DEFAULT_SETTING_PATH = "classpath:nacos_default_setting.properties";
```
Comment (KomachiSion) [inline client/src/main/java/com/alibaba/nacos/client/env/DefaultSettingPropertySource.java:31]: I only found the an test `nacos_default_setting.properties`, do you want to add a real `nacos_default_setting.properties` ?

**Review by KomachiSion — CHANGES_REQUESTED** [review]

**Review by KomachiSion — APPROVED** [review]

**Pull request merged**